In [509]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [510]:
names = ['preg','plasmaglu', 'bloodpressure','sknthckness', 'insulin', 'bmi', 'diabetes', 'age', 'class']

In [511]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/pima-indians-diabetes/pima-indians-diabetes.data', names=names)

In [512]:
for feature in ['plasmaglu','bloodpressure','sknthckness','insulin','bmi']:
   df.loc[:,feature].replace(to_replace=0, value=np.nan, inplace=True)

In [513]:
for f in ['plasmaglu','bmi','bloodpressure','insulin','sknthckness']:
  med = df[f].median()
  df[f].fillna(med, inplace=True)

In [514]:
df_labels = df['diabetes']
df_X = df.drop('diabetes', axis=1)

In [515]:
y, X = df['class'], df.drop('class', axis=1)

In [516]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [558]:
from sklearn.preprocessing import StandardScaler
stdsc = StandardScaler()
X_train_std = stdsc.fit_transform(X_train)
X_test_std = stdsc.transform(X_test)

In [559]:
X_train_std.shape

(614, 8)

In [560]:
from sklearn.cross_validation import ShuffleSplit

In [561]:
from sklearn.grid_search import GridSearchCV

In [565]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()

from sklearn.cross_validation import ShuffleSplit
cv = ShuffleSplit(X_train_std.shape[0], n_iter=10, test_size=0.2, random_state=0)

from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=1, weights='uniform', p=2)
knn.fit(X_train_std,y_train) 

from sklearn.grid_search import GridSearchCV

parameters = {'n_neighbors': range(1,30), 'weights': ['uniform', 'distance'],
             'p': range(1,4)}
clf = GridSearchCV(knn, parameters, cv=cv)
clf.fit(X_train_std, y_train)


GridSearchCV(cv=ShuffleSplit(614, n_iter=10, test_size=0.2, random_state=0),
       error_score='raise',
       estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_neighbors=1, p=2, weights='uniform'),
       fit_params={}, iid=True, loss_func=None, n_jobs=1,
       param_grid={'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29], 'weights': ['uniform', 'distance'], 'p': [1, 2, 3]},
       pre_dispatch='2*n_jobs', refit=True, score_func=None, scoring=None,
       verbose=0)

In [566]:
clf.best_score_

0.22357723577235772

In [522]:
from sklearn.learning_curve import learning_curve
title = 'title'

In [523]:
from sklearn.linear_model import LinearRegression

In [524]:
lr = LinearRegression()
lr.fit(X_train_std, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [525]:
from sklearn.cross_validation import ShuffleSplit
cv = ShuffleSplit(X_train.shape[0], n_iter=10, test_size=0.2, random_state=0)

In [526]:
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression(penalty='l2', C=10)

In [527]:
logreg.fit(X_train_std, y_train)
zip(feature_cols, logreg.coef_[0])

[('preg', 0.37684432401936041), ('insulin', 1.0661971697635391)]

In [528]:
y_pred_class = logreg.predict(X_test_std)

In [529]:
from sklearn import metrics
print metrics.accuracy_score(y_test, y_pred_class)

0.776041666667


In [530]:
import pandas as pd
%matplotlib inline

In [531]:
from sklearn.naive_bayes import GaussianNB
from sklearn.grid_search import GridSearchCV
from sklearn.cross_validation import KFold, cross_val_score

In [567]:
y, X = df['preg'], df.drop(['preg'], axis=1)

In [568]:
nb = GaussianNB()
nb.fit(X_train_std, y_train)

GaussianNB()

In [576]:
nb.score(X_train_std, y_train)

0.23127035830618892

In [577]:
import pandas as pd
import json

In [578]:
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier()

In [579]:
from sklearn.cross_validation import cross_val_score

In [580]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators = 20)
    
model.fit(X_train_std, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=20, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [581]:
scores = cross_val_score(model, X_train_std, y_train, scoring='accuracy', cv=5)
print('CV accuracy {}, Average accuracy {}'.format(scores, scores.mean()))

CV accuracy [ 0.19230769  0.2519685   0.23577236  0.1965812   0.16239316], Average accuracy 0.207804582589


//anaconda/lib/python2.7/site-packages/sklearn/cross_validation.py:417: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of labels for any class cannot be less than n_folds=5.
  % (min_labels, self.n_folds)), Warning)


In [582]:
from sklearn.cross_validation import cross_val_score

scores = cross_val_score(model, X_train_std, y_train, scoring='accuracy')
print('CV accuracy {}, Average accuracy {}'.format(scores, scores.mean()))

for n_trees in range(1, 100, 10):
    model = RandomForestClassifier(n_estimators = n_trees)
    scores = cross_val_score(model, X_train_std, y_train, scoring='accuracy')
    print('CV accuracy {}, Average accuracy {}'.format(scores, scores.mean()))

CV accuracy [ 0.1952381   0.19902913  0.1969697 ], Average accuracy 0.197078972807
CV accuracy [ 0.2047619   0.14563107  0.13131313], Average accuracy 0.160568701345
CV accuracy [ 0.18095238  0.16504854  0.1969697 ], Average accuracy 0.180990207204
CV accuracy [ 0.16190476  0.18932039  0.17676768], Average accuracy 0.175997609007
CV accuracy [ 0.20952381  0.17961165  0.16666667], Average accuracy 0.185267375559
CV accuracy [ 0.21904762  0.17961165  0.20707071], Average accuracy 0.201909992201
CV accuracy [ 0.21428571  0.21359223  0.16161616], Average accuracy 0.196498036304
CV accuracy [ 0.21428571  0.20873786  0.18181818], Average accuracy 0.201613920061
CV accuracy [ 0.21428571  0.19902913  0.20707071], Average accuracy 0.206795182523
CV accuracy [ 0.22380952  0.2184466   0.22222222], Average accuracy 0.221492782658
CV accuracy [ 0.17619048  0.17961165  0.14646465], Average accuracy 0.167422257714


//anaconda/lib/python2.7/site-packages/sklearn/cross_validation.py:417: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of labels for any class cannot be less than n_folds=3.
  % (min_labels, self.n_folds)), Warning)


# Describe the content of the dataset and its goals
This dataset looks at the Pima Indian female population. They have a high incidence rate of diabetes. If the Indian Health Service Hospital found a glucose concentration of at least 200 mg/dl the patient would be diagnosed with diabetes. Eight variables were chosen for predicting the onset of diabets in Pima Indian women. These variables include # of times pregnent, plasma glucose concentration at 2 hours in a test, diastolic blood pressire, skin fold thickness, 2-hour serum insulin, BMI, diabets pedigree function and age.


# Describe the features and formulate a hypothesis on which might be relevant in predicting diabetes

I predict woman with plasma glucose concentrations of at least 200 mg/dl have a higher chance of having diabetes. Pregnant woman also experience high blood glucose levels.

# Describe the missing/NULL values. Decide if you should impute or drop them and justify your choice.

The dataset states that there are zeros in places where they are biologically impossible, such as gluclose level, BMI, blood pressure, insulin level, and skin thickness. As a result, I dropped lines that have these values. 

Just using pregnancy and 2 hour serum insulin to fit class with Logistic Regression, my model was 77% accurate. 

Using Gaussian Naive Bayes my model was 23% accurate. 

Using KKN my model was 22% accurate. 

Using a Decision Tree my model was 20% accurate. 

I have learned that I first needed to define x and y, take out 0s, and then place all of my modeling after that.

